src/crdt_rga.erl

```diff
@@ -0,0 +1,288 @@
+%% @doc
+%%
+%% An operation-based Replicated Growable Array CRDT.
+%%
+%% As the data structure is operation-based, to issue an operation, one should
+%% first call `generate_downstream/3', to get the downstream version of the
+%% operation, and then call `update/2'.
+%%
+%% It provides two operations: addRight, which adds an element to the RGA, and remove,
+%% which removes an element from the RGA.
+%%
+%% This implementation is based on the paper cited below.
+%%
+%% @reference Marc Shapiro, Nuno Preguiça, Carlos Baquero, Marek Zawirski (2011) A comprehensive study of
+%% Convergent and Commutative Replicated Data Types. http://hal.upmc.fr/inria-00555588/
+%%
+%% @end
+-module(crdt_rga).
+
+-export([new/0, update/2, purge_tombstones/1, generate_downstream/3, value/1, is_operation/1]).
+
+-export_type([rga/0, rga_op/0, rga_downstream_op/0]).
+
+-ifdef(TEST).
+-include_lib("eunit/include/eunit.hrl").
+-endif.
+
+-type vertex() :: {atom(), any(), number()}.
+
+-type rga_downstream_op() :: {addRight, vertex(), vertex()} | {remove, vertex()}.
+
+-type rga_op() :: {addRight, any(), number()} | {remove, number()}.
+
+-type rga_result() :: {ok, rga()}.
+
+-type rga() :: list().
+
+-type actor() :: riak_dt:actor().
+
+-spec new() -> [].
+new() ->
+    [].
+
+%% @doc generate downstream operations.
+%% If the operation is addRight, generates a unique token for the new element.
+%% If the operation is remove, fetches the vertex of the element to be removed.
+-spec generate_downstream(rga_op(), actor(), rga()) -> {ok, rga_downstream_op()} | {error, {invalid_position, number()}}.
+generate_downstream({addRight, Elem, Position}, _Actor, Rga) ->
+    case (Position < 0) or (Position > length(Rga)) of
+        true -> {error, {invalid_position, Position}};
+        false -> case length(Rga) of
+                     0 -> {ok, {addRight, {ok, 0, 0}, {ok, Elem, unique()}}};
+                     _ -> {ok, {addRight, lists:nth(Position, Rga), {ok, Elem, unique()}}}
+                 end
+    end;
+generate_downstream({remove, Position}, _Actor, Rga) ->
+    {ok, {remove, lists:nth(Position, Rga)}}.
+
+%% @doc given an RGA, returns the same RGA, as it represents its own value.
+-spec value(rga()) -> rga().
+value(Rga) ->
+    Rga.
+
+%% @doc This method takes in an rga operation and an rga to perform the operation.
+%% It returns either the added Vertex and the new rga in case of an addRight, and the new rga in the case of a remove.
+-spec update(rga_downstream_op(), rga()) -> rga_result().
+update({addRight, RightVertex, NewVertex}, Rga) ->
+    recursive_insert(RightVertex, NewVertex, Rga, []);
+update({remove, Vertex}, Rga) ->
+    recursive_remove(Vertex, Rga, []).
+
+%% Private
+%% @doc recursively looks for the Vertex where the new element should be put to the right of.
+-spec recursive_insert(vertex(), vertex(), rga(), list()) -> rga_result().
```

Comment:
Annette, according to the documentation of the method lists:take_while: "returns the longest prefix of the list for which all elements satisfy the predicate", but I also need the other elements which come after. In order to do it like this, I think I would also need to use lists:substract and lists:merge to get the result I want. I don't know if its better to do it this way regarding complexity. What do you think?